File: steps/javascript/adoption.summary.step.js
// steps/javascript/adoption.summary.step.js
const { get } = require('./js-store');
const { PetRecommendationAgent } = require('./agent');

exports.config = {
  type: 'event',
  name: 'JsApplicationSummarizer',
  description: 'AI agent that summarizes application text in parallel with background check',
  subscribes: ['js.adoption.applied'],
  emits: ['js.adoption.summary.complete'],
  flows: ['javascript-adoptions'],
};

exports.handler = async (event, { emit, logger, streams, traceId }) => {
  const { applicationId, petId, adopterName, adopterEmail } = event || {};

  logger.info('📝 Generating application summary', { applicationId, petId, adopterName });

  try {
    // Get pet information for context
    const pet = get(petId);
    const petName = pet ? pet.name : 'Unknown Pet';
    const petSpecies = pet ? pet.species : 'unknown';

    // Create application data for AI summarization
    const applicationData = {
      applicationId,
      petName,
      petSpecies,
      adopterName: adopterName || 'Unknown Adopter',
      adopterEmail: adopterEmail || 'No email provided',
      checkResult: 'pending' // Will be updated when background check completes
    };

    // Generate AI-powered summary
    const summary = await PetRecommendationAgent.generateApplicationSummary(applicationData);

    logger.info('Application summary generated', { 
      applicationId, 
      petId, 
      summary: summary.substring(0, 100) + '...' 
    });

    // Update stream with summary
    if (streams?.adoptions && traceId) {
      await streams.adoptions.set(traceId, 'summary', {
        entityId: applicationId,
        type: 'application',
        phase: 'summary_ready',
        message: summary,
        timestamp: Date.now(),
        data: { petId, petName, adopterName }
      });
    }

    // Emit completion event with summary
    await emit({
      topic: 'js.adoption.summary.complete',
      data: {
        applicationId,
        petId,
        petName,
        adopterName,
        adopterEmail,
        summary,
        traceId
      }
    });

  } catch (error) {
    logger.error('Application summary generation failed', { 
      applicationId, 
      petId, 
      error: error.message 
    });

    // Fallback summary
    const fallbackSummary = `Application ${applicationId} for ${adopterName || 'adopter'} requires review.`;

    await emit({
      topic: 'js.adoption.summary.complete',
      data: {
        applicationId,
        petId,
        adopterName,
        adopterEmail,
        summary: fallbackSummary,
        error: error.message,
        traceId
      }
    });
  }
};

File: steps/python/adoption_summary.step.py
# steps/python/adoption_summary.step.py
import sys
import os
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))

from services import pet_store
from agent import PetRecommendationAgent
import time

config = {
    "type": "event",
    "name": "PyApplicationSummarizer",
    "description": "AI agent that summarizes application text in parallel with background check",
    "subscribes": ["py.adoption.applied"],
    "emits": ["py.adoption.summary.complete"],
    "flows": ["python-adoptions"]
}

async def handler(event, ctx=None):
    if not ctx:
        return
        
    application_id = (event or {}).get("applicationId")
    pet_id = (event or {}).get("petId")
    adopter_name = (event or {}).get("adopterName")
    adopter_email = (event or {}).get("adopterEmail")
    
    ctx.logger.info("📝 Generating application summary", {
        "applicationId": application_id,
        "petId": pet_id,
        "adopterName": adopter_name
    })

    try:
        # Get pet information for context
        pet = pet_store.get(pet_id)
        pet_name = pet.get("name", "Unknown Pet") if pet else "Unknown Pet"
        pet_species = pet.get("species", "unknown") if pet else "unknown"

        # Create application data for AI summarization
        application_data = {
            "applicationId": application_id,
            "petName": pet_name,
            "petSpecies": pet_species,
            "adopterName": adopter_name or "Unknown Adopter",
            "adopterEmail": adopter_email or "No email provided",
            "checkResult": "pending"  # Will be updated when background check completes
        }

        # Generate AI-powered summary
        summary = await PetRecommendationAgent.generate_application_summary(application_data)

        ctx.logger.info("Application summary generated", {
            "applicationId": application_id,
            "petId": pet_id,
            "summary": summary[:100] + "..." if len(summary) > 100 else summary
        })

        # Update stream with summary
        # Summary generated - no stream update needed

        # Emit completion event with summary
        await ctx.emit({
            "topic": "py.adoption.summary.complete",
            "data": {
                "applicationId": application_id,
                "petId": pet_id,
                "petName": pet_name,
                "adopterName": adopter_name,
                "adopterEmail": adopter_email,
                "summary": summary
            }
        })

    except Exception as error:
        ctx.logger.error("Application summary generation failed", {
            "applicationId": application_id,
            "petId": pet_id,
            "error": str(error)
        })

        # Fallback summary
        fallback_summary = f"Application {application_id} for {adopter_name or 'adopter'} requires review."

        await ctx.emit({
            "topic": "py.adoption.summary.complete",
            "data": {
                "applicationId": application_id,
                "petId": pet_id,
                "adopterName": adopter_name,
                "adopterEmail": adopter_email,
                "summary": fallback_summary,
                "error": str(error)
            }
        })
